File: packages/language-expression/src/stdlib/collections/ReadonlyCollection.ts
import { classType, typeRef, lambdaType, genericTypeRef } from "../../typir-extensions/config/typeBuilder.js";

/**
 * The built-in generic ReadonlyCollection type exported as `ReadonlyCollectionType`.
 */
export const ReadonlyCollectionType = classType("ReadonlyCollection", "builtin")
    .generics("T")
    .extends("Any")
    .method("asBag", (m) => m.signature((s) => s.returns(typeRef("void").build())))
    .method("asOrderedSet", (m) => m.signature((s) => s.returns(typeRef("void").build())))
    .method("asList", (m) => m.signature((s) => s.returns(typeRef("void").build())))
    .method("asSet", (m) => m.signature((s) => s.returns(typeRef("void").build())))
    .method("clone", (m) =>
        m.signature((s) =>
            s.returns(
                typeRef("Collection")
                    .withTypeArgs(new Map([["T", genericTypeRef("T")]]))
                    .build()
            )
        )
    )
    .method("concat", (m) =>
        m
            .signature((s) => s.returns(typeRef("string").build()))
            .signature((s) => s.param("separator", typeRef("string").build()).returns(typeRef("string").build()))
    )
    .method("count", (m) =>
        m
            .signature((s) => s.param("item", typeRef("Any").build()).returns(typeRef("int").build()))
            .signature((s) =>
                s
                    .param(
                        "iterator",
                        lambdaType().param("it", genericTypeRef("T")).returns(typeRef("boolean").build())
                    )
                    .returns(typeRef("int").build())
            )
    )
    .method("excludes", (m) =>
        m.signature((s) => s.param("item", typeRef("Any").build()).returns(typeRef("boolean").build()))
    )
    .method("excludesAll", (m) =>
        m.signature((s) =>
            s
                .param(
                    "col",
                    typeRef("Collection")
                        .withTypeArgs(new Map([["T", typeRef("Any").build()]]))
                        .build()
                )
                .returns(typeRef("boolean").build())
        )
    )
    .method("excluding", (m) =>
        m.signature((s) =>
            s.param("item", typeRef("Any").build()).returns(
                typeRef("Collection")
                    .withTypeArgs(new Map([["T", genericTypeRef("T")]]))
                    .build()
            )
        )
    )
    .method("excludingAll", (m) =>
        m.signature((s) =>
            s
                .param(
                    "col",
                    typeRef("Collection")
                        .withTypeArgs(new Map([["T", typeRef("Any").build()]]))
                        .build()
                )
                .returns(
                    typeRef("Collection")
                        .withTypeArgs(new Map([["T", genericTypeRef("T")]]))
                        .build()
                )
        )
    )
    .method("flatten", (m) =>
        m.signature((s) =>
            s.returns(
                typeRef("Collection")
                    .withTypeArgs(new Map([["T", typeRef("Any").build()]]))
                    .build()
            )
        )
    )
    .method("includes", (m) =>
        m.signature((s) => s.param("item", typeRef("Any").build()).returns(typeRef("boolean").build()))
    )
    .method("includesAll", (m) =>
        m.signature((s) =>
            s
                .param(
                    "col",
                    typeRef("Collection")
                        .withTypeArgs(new Map([["T", typeRef("Any").build()]]))
                        .build()
                )
                .returns(typeRef("boolean").build())
        )
    )
    .method("including", (m) =>
        m.signature((s) =>
            s.param("item", typeRef("Any").build()).returns(
                typeRef("Collection")
                    .withTypeArgs(new Map([["T", genericTypeRef("T")]]))
                    .build()
            )
        )
    )
    .method("includingAll", (m) =>
        m.signature((s) =>
            s
                .param(
                    "col",
                    typeRef("Collection")
                        .withTypeArgs(new Map([["T", typeRef("Any").build()]]))
                        .build()
                )
                .returns(
                    typeRef("Collection")
                        .withTypeArgs(new Map([["T", genericTypeRef("T")]]))
                        .build()
                )
        )
    )
    .method("isEmpty", (m) => m.signature((s) => s.returns(typeRef("boolean").build())))
    .method("min", (m) =>
        m
            .signature((s) => s.returns(typeRef("double").nullable().build()))
            .signature((s) => s.param("default", typeRef("double").build()).returns(typeRef("double").build()))
    )
    .method("max", (m) =>
        m
            .signature((s) => s.returns(typeRef("double").nullable().build()))
            .signature((s) => s.param("default", typeRef("double").build()).returns(typeRef("double").build()))
    )
    .method("notEmpty", (m) => m.signature((s) => s.returns(typeRef("boolean").build())))
    .method("random", (m) => m.signature((s) => s.returns(typeRef("Any").build())))
    .method("size", (m) => m.signature((s) => s.returns(typeRef("int").build())))
    .method("sum", (m) => m.signature((s) => s.returns(typeRef("double").build())))
    .method("atLeastNMatch", (m) =>
        m.signature((s) =>
            s
                .param("iterator", lambdaType().param("it", genericTypeRef("T")).returns(typeRef("boolean").build()))
                .param("n", typeRef("int").build())
                .returns(typeRef("boolean").build())
        )
    )
    .method("atMostNMatch", (m) =>
        m.signature((s) =>
            s
                .param("iterator", lambdaType().param("it", genericTypeRef("T")).returns(typeRef("boolean").build()))
                .param("n", typeRef("int").build())
                .returns(typeRef("boolean").build())
        )
    )
    .method("aggregate", (m) =>
        m.signature((s) =>
            s.param("iterator", lambdaType().param("it", genericTypeRef("T")).returns(typeRef("Any").build())).returns(
                typeRef("Map")
                    .withTypeArgs(
                        new Map([
                            ["K", typeRef("Any").build()],
                            ["V", typeRef("Any").build()]
                        ])
                    )
                    .build()
            )
        )
    )
    .method("map", (m) =>
        m.signature((s) =>
            s
                .generics("U")
                .param("iterator", lambdaType().param("it", genericTypeRef("T")).returns(genericTypeRef("U")))
                .returns(
                    typeRef("ReadonlyCollection")
                        .withTypeArgs(new Map([["T", genericTypeRef("U")]]))
                        .build()
                )
        )
    )
    .method("exists", (m) =>
        m.signature((s) =>
            s
                .param("iterator", lambdaType().param("it", genericTypeRef("T")).returns(typeRef("boolean").build()))
                .returns(typeRef("boolean").build())
        )
    )
    .method("forAll", (m) =>
        m.signature((s) =>
            s
                .param("iterator", lambdaType().param("it", genericTypeRef("T")).returns(typeRef("boolean").build()))
                .returns(typeRef("boolean").build())
        )
    )
    .method("associate", (m) =>
        m.signature((s) =>
            s
                .generics("U")
                .param("iterator", lambdaType().param("it", genericTypeRef("T")).returns(genericTypeRef("U")))
                .returns(
                    typeRef("ReadonlyMap")
                        .withTypeArgs(
                            new Map([
                                ["K", genericTypeRef("T")],
                                ["V", genericTypeRef("U")]
                            ])
                        )
                        .build()
                )
        )
    )
    .method("nMatch", (m) =>
        m.signature((s) =>
            s
                .param("iterator", lambdaType().param("it", genericTypeRef("T")).returns(typeRef("boolean").build()))
                .param("n", typeRef("int").build())
                .returns(typeRef("boolean").build())
        )
    )
    .method("none", (m) =>
        m.signature((s) =>
            s
                .param("iterator", lambdaType().param("it", genericTypeRef("T")).returns(typeRef("boolean").build()))
                .returns(typeRef("boolean").build())
        )
    )
    .method("one", (m) =>
        m.signature((s) =>
            s
                .param("iterator", lambdaType().param("it", genericTypeRef("T")).returns(typeRef("boolean").build()))
                .returns(typeRef("boolean").build())
        )
    )
    .method("reject", (m) =>
        m.signature((s) =>
            s
                .param("iterator", lambdaType().param("it", genericTypeRef("T")).returns(typeRef("boolean").build()))
                .returns(
                    typeRef("Collection")
                        .withTypeArgs(new Map([["T", genericTypeRef("T")]]))
                        .build()
                )
        )
    )
    .method("rejectOne", (m) =>
        m.signature((s) =>
            s
                .param("iterator", lambdaType().param("it", genericTypeRef("T")).returns(typeRef("boolean").build()))
                .returns(
                    typeRef("Collection")
                        .withTypeArgs(new Map([["T", genericTypeRef("T")]]))
                        .build()
                )
        )
    )
    .method("filter", (m) =>
        m.signature((s) =>
            s
                .param("iterator", lambdaType().param("it", genericTypeRef("T")).returns(typeRef("boolean").build()))
                .returns(
                    typeRef("Collection")
                        .withTypeArgs(new Map([["T", genericTypeRef("T")]]))
                        .build()
                )
        )
    )
    .method("filterByKind", (m) =>
        m.signature((s) =>
            s.returns(
                typeRef("Collection")
                    .withTypeArgs(new Map([["T", typeRef("Any").build()]]))
                    .build()
            )
        )
    )
    .method("filterByType", (m) =>
        m.signature((s) =>
            s.returns(
                typeRef("Collection")
                    .withTypeArgs(new Map([["T", typeRef("Any").build()]]))
                    .build()
            )
        )
    )
    .method("find", (m) =>
        m.signature((s) =>
            s
                .param("iterator", lambdaType().param("it", genericTypeRef("T")).returns(typeRef("boolean").build()))
                .returns(typeRef("Any").build())
        )
    )
    .method("sortedBy", (m) =>
        m.signature((s) =>
            s
                .generics("U")
                .param("iterator", lambdaType().param("it", genericTypeRef("T")).returns(genericTypeRef("U")))
                .returns(
                    typeRef("ReadonlyOrderedCollection")
                        .withTypeArgs(new Map([["T", genericTypeRef("T")]]))
                        .build()
                )
        )
    )
    .build();

File: platform/script/src/main/kotlin/com/mdeo/script/stdlib/collections/ReadonlyCollection.kt
package com.mdeo.script.stdlib.collections

import java.util.function.Function
import java.util.function.Predicate

/**
 * A readonly collection that provides query operations without modification capabilities.
 * This interface follows OCL/EOL collection semantics.
 *
 * @param T the type of elements contained in this collection
 */
interface ReadonlyCollection<out T> : ScriptIterable<T> {

    /**
     * Returns the number of elements in this collection.
     */
    fun size(): Int

    /**
     * Returns true if this collection contains no elements.
     */
    fun isEmpty(): Boolean

    /**
     * Returns true if this collection contains at least one element.
     */
    fun notEmpty(): Boolean

    /**
     * Returns true if this collection contains the specified element.
     *
     * @param item the element to check for
     */
    fun includes(item: Any?): Boolean

    /**
     * Returns true if this collection does not contain the specified element.
     *
     * @param item the element to check for
     */
    fun excludes(item: Any?): Boolean

    /**
     * Returns true if this collection contains all elements from the given collection.
     *
     * @param col the collection of elements to check for
     */
    fun includesAll(col: ReadonlyCollection<@UnsafeVariance T>): Boolean

    /**
     * Returns true if this collection contains none of the elements from the given collection.
     *
     * @param col the collection of elements to check for
     */
    fun excludesAll(col: ReadonlyCollection<@UnsafeVariance T>): Boolean

    /**
     * Returns the count of occurrences of the specified item in this collection.
     *
     * @param item the element to count
     */
    fun count(item: Any?): Int

    /**
     * Returns the count of elements that match the given predicate.
     *
     * @param predicate the predicate to apply to each element
     */
    fun count(predicate: Predicate<@UnsafeVariance T>): Int

    /**
     * Returns a new collection containing all elements except the specified item.
     *
     * @param item the element to exclude
     */
    fun excluding(item: Any?): ReadonlyCollection<T>

    /**
     * Returns a new collection containing all elements except those in the given collection.
     *
     * @param col the collection of elements to exclude
     */
    fun excludingAll(col: ReadonlyCollection<@UnsafeVariance T>): ReadonlyCollection<T>

    /**
     * Returns a new collection containing all elements plus the specified item.
     *
     * @param item the element to include
     */
    fun including(item: @UnsafeVariance T): ReadonlyCollection<T>

    /**
     * Returns a new collection containing all elements plus those from the given collection.
     *
     * @param col the collection of elements to include
     */
    fun includingAll(col: ReadonlyCollection<@UnsafeVariance T>): ReadonlyCollection<T>

    /**
     * Returns a random element from this collection.
     *
     * @throws NoSuchElementException if the collection is empty
     */
    fun random(): T

    /**
     * Returns the sum of all elements (elements must be numeric).
     */
    fun sum(): Double

    /**
     * Concatenates all elements to a string without separator.
     */
    fun concat(): String

    /**
     * Concatenates all elements to a string with the given separator.
     *
     * @param separator the separator to use between elements
     */
    fun concat(separator: String): String

    /**
     * Flattens nested collections into a single collection.
     */
    fun flatten(): ReadonlyCollection<Any?>

    /**
     * Converts this collection to a Bag.
     */
    fun asBag(): Bag<@UnsafeVariance T>

    /**
     * Converts this collection to an OrderedSet.
     */
    fun asOrderedSet(): OrderedSet<@UnsafeVariance T>

    /**
     * Converts this collection to a List.
     */
    fun asList(): ScriptList<@UnsafeVariance T>

    /**
     * Converts this collection to a Set.
     */
    fun asSet(): ScriptSet<@UnsafeVariance T>

    /**
     * Creates a clone of this collection.
     */
    fun clone(): Collection<@UnsafeVariance T>

    /**
     * Returns true if at least n elements match the given predicate.
     *
     * @param predicate the predicate to apply
     * @param n the minimum number of matches required
     */
    fun atLeastNMatch(predicate: Predicate<@UnsafeVariance T>, n: Int): Boolean

    /**
     * Returns true if at most n elements match the given predicate.
     *
     * @param predicate the predicate to apply
     * @param n the maximum number of matches allowed
     */
    fun atMostNMatch(predicate: Predicate<@UnsafeVariance T>, n: Int): Boolean

    /**
     * Groups elements by a key function.
     *
     * @param keyMapper the function to extract the key from each element
     */
    fun aggregate(keyMapper: Function<@UnsafeVariance T, Any?>): ScriptMap<Any?, ScriptList<@UnsafeVariance T>>

    /**
     * Maps each element to a new value.
     *
     * @param mapper the mapping function
     */
    fun <U> map(mapper: Function<@UnsafeVariance T, U>): ReadonlyCollection<U>

    /**
     * Returns true if at least one element matches the predicate.
     *
     * @param predicate the predicate to apply
     */
    fun exists(predicate: Predicate<@UnsafeVariance T>): Boolean

    /**
     * Returns true if all elements match the predicate.
     *
     * @param predicate the predicate to apply
     */
    fun forAll(predicate: Predicate<@UnsafeVariance T>): Boolean

    /**
     * Creates a map by associating each element with a computed value.
     *
     * @param valueMapper the function to compute the value for each element
     */
    fun <U> associate(valueMapper: Function<@UnsafeVariance T, U>): ReadonlyMap<@UnsafeVariance T, U>

    /**
     * Returns true if exactly n elements match the predicate.
     *
     * @param predicate the predicate to apply
     * @param n the exact number of matches required
     */
    fun nMatch(predicate: Predicate<@UnsafeVariance T>, n: Int): Boolean

    /**
     * Returns true if no elements match the predicate.
     *
     * @param predicate the predicate to apply
     */
    fun none(predicate: Predicate<@UnsafeVariance T>): Boolean

    /**
     * Returns true if exactly one element matches the predicate.
     *
     * @param predicate the predicate to apply
     */
    fun one(predicate: Predicate<@UnsafeVariance T>): Boolean

    /**
     * Returns a new collection containing only elements that don't match the predicate.
     *
     * @param predicate the predicate to apply
     */
    fun reject(predicate: Predicate<@UnsafeVariance T>): Collection<@UnsafeVariance T>

    /**
     * Returns a new collection with the first matching element removed.
     *
     * @param predicate the predicate to apply
     */
    fun rejectOne(predicate: Predicate<@UnsafeVariance T>): Collection<@UnsafeVariance T>

    /**
     * Returns a new collection containing only elements that match the predicate.
     * Alias for select in OCL.
     *
     * @param predicate the predicate to apply
     */
    fun filter(predicate: Predicate<@UnsafeVariance T>): Collection<@UnsafeVariance T>

    /**
     * Finds the first element matching the predicate.
     *
     * @param predicate the predicate to apply
     * @return the first matching element or null if not found
     */
    fun find(predicate: Predicate<@UnsafeVariance T>): T?

    /**
     * Returns an ordered collection sorted by the given key extractor.
     *
     * @param keyExtractor the function to extract the sort key
     */
    fun <U : Comparable<U>> sortedBy(keyExtractor: Function<@UnsafeVariance T, U>): ReadonlyOrderedCollection<@UnsafeVariance T>
}
